File: scripts/python/src/tools/dotfile/secret/store.py
import os
import subprocess

from tools.core.paths import tilde
from tools.core.process import run
from tools.dotfile.secret.apply import prepare
from tools.dotfile.secret.identity import sops_env
from tools.dotfile.secret.keys import load_recipients
from tools.dotfile.secret.variables import VARS, references, vars_file
from tools.dotfile.secret.variables import load as load_vars
from tools.dotfile.secret.vault import (
    ENC,
    FILE_MODE,
    MARKER,
    SUFFIX,
    TMPL,
    Entry,
    encrypt,
    encrypt_text,
    have_key,
    materialise,
    plan,
)
from tools.dotfile.state import canon, die, log, shorten
from tools.dotfile.system import plan as system_plan
from tools.dotfile.targets import load_targets

UNCHANGED = 200
EMPTY_MAPPING = "{}\n"


def git_add(ctx, *paths):
    subprocess.run(
        ["git", "-C", ctx.root, "add", *paths],
        stderr=subprocess.DEVNULL,
        check=False,
    )


def targets_has_line(ctx, line):
    try:
        with open(ctx.targets_file, encoding="utf-8") as handle:
            return line in handle.read().splitlines()
    except OSError:
        return False


def require_vault(ctx):
    if not load_recipients(ctx):
        die("no recipients enrolled (run: dotfile secret enroll <label>)")
    if not have_key(ctx):
        die("no age identity on this machine (run: dotfile secret init)")


def resolve_source(ctx, path):
    expanded = ctx.home + path[1:] if path.startswith("~") else path
    if not expanded.startswith("/"):
        expanded = os.path.join(os.getcwd(), expanded)
    src = canon(expanded)
    if os.path.islink(src):
        die(f"refusing to adopt a symlink: {shorten(ctx, src)}")
    if not os.path.isfile(src):
        die(f"not a file: {shorten(ctx, src)}")
    if not src.startswith(ctx.home + "/"):
        die("source must live under $HOME")
    return src


def plan_destination(ctx, src, group, pkg):
    parent = os.path.dirname(src)
    destrel = f"{group}/{pkg}/{os.path.basename(src)}{SUFFIX}"
    default = os.path.join(ctx.home, ".config", pkg)
    mapline = "" if parent == default else f"{group}/{pkg} = {tilde(parent)}"
    return destrel, mapline


def cmd_add(ctx, path, group, pkg, marker):
    if not pkg:
        die("--pkg <name> is required")
    require_vault(ctx)
    src = resolve_source(ctx, path)
    load_targets(ctx)

    destrel, mapline = plan_destination(ctx, src, group, pkg)
    dest = os.path.join(ctx.root, destrel)
    if os.path.exists(dest):
        die(f"destination exists: {destrel}")

    pkgdir = os.path.dirname(dest)
    fresh = not os.path.isdir(pkgdir)

    problem = encrypt(ctx, src, dest)
    if problem:
        die(problem)
    log(f"sealed {shorten(ctx, src)} -> {destrel}")

    marker_path = os.path.join(pkgdir, MARKER)
    if marker is not False and (marker or fresh) and not os.path.exists(marker_path):
        with open(marker_path, "w", encoding="utf-8"):
            pass
        log(f"marked {group}/{pkg} as a secret package")
        git_add(ctx, marker_path)

    os.chmod(src, FILE_MODE)
    log(f"kept   {shorten(ctx, src)} (0600)")

    if mapline and not targets_has_line(ctx, mapline):
        with open(ctx.targets_file, "a", encoding="utf-8") as handle:
            handle.write(mapline + "\n")
        log(f"mapped {mapline}")

    git_add(ctx, dest, ctx.targets_file)


def cmd_vars(ctx, unused_only):
    prepare(ctx)
    declared = load_vars(ctx)
    if declared.note:
        die(declared.note)
    if not declared.values:
        log(f"nothing declared in {VARS}")
        return
    used = {}
    for entry in plan(ctx) + system_plan(ctx):
        if entry.kind != TMPL:
            continue
        with open(entry.src, encoding="utf-8", errors="replace") as handle:
            for name in references(handle.read()):
                used.setdefault(name, []).append(shorten(ctx, entry.dst))
    names = sorted(declared.values)
    if unused_only:
        names = [name for name in names if name not in used]
    if not names:
        log("every var is referenced")
        return
    width = max(len(name) for name in names)
    for name in names:
        where = " ".join(used.get(name, [])) or "unused"
        log(f"  {name:<{width}}  {where}")
    missing = sorted(set(used) - set(declared.values))
    for name in missing:
        log(f"  {name}  referenced by {' '.join(used[name])} but not defined")
    if missing:
        raise SystemExit(1)


def matching_entries(ctx, path):
    expanded = ctx.home + path[1:] if path.startswith("~") else path
    declared_path = vars_file(ctx)
    if expanded == declared_path or path in (VARS, "vars"):
        return [Entry(declared_path, "", VARS, ENC)]
    entries = plan(ctx)
    exact = [entry for entry in entries if entry.dst == expanded or entry.src == expanded]
    if exact:
        return exact
    return [
        entry
        for entry in entries
        if entry.src.endswith("/" + path.lstrip("/")) or entry.dst.endswith("/" + path.lstrip("/"))
    ]


def cmd_edit(ctx, path):
    require_vault(ctx)
    prepare(ctx)
    found = matching_entries(ctx, path)
    if not found:
        die(f"no tracked secret matches '{path}'")
    if len(found) > 1:
        names = " ".join(entry.src[len(ctx.root) + 1 :] for entry in found)
        die(f"'{path}' matches several: {names}")

    entry = found[0]
    fresh = not os.path.exists(entry.src)
    if fresh:
        seed_vars(ctx, entry.src)
    result = run(["sops", entry.src], cwd=ctx.root, env=sops_env(ctx))
    if result.returncode == UNCHANGED:
        log(f"unchanged {shorten(ctx, entry.src)}")
        git_add(ctx, entry.src)
        return
    if result.returncode != 0:
        if fresh:
            os.remove(entry.src)
        raise SystemExit(result.returncode)

    git_add(ctx, entry.src)
    if not entry.dst:
        log(f"saved {entry.src[len(ctx.root) + 1 :]}")
        return
    state = materialise(ctx, entry, load_vars(ctx), False, True)
    log(f"{state} {shorten(ctx, entry.dst)}")


def seed_vars(ctx, path):
    if os.path.basename(path) != VARS:
        die(f"nothing to edit: {shorten(ctx, path)}")
    problem = encrypt_text(ctx, path, EMPTY_MAPPING)
    if problem:
        die(problem)

File: scripts/rust/crates/dotfile/src/secret/store.rs
use super::{cli::AddArgs, recipients, sops, vault};
use crate::context::Context;
use std::collections::{BTreeMap, BTreeSet};
use std::fs;
use std::path::Path;
use std::process::ExitCode;

fn require_vault(context: &Context) -> Result<(), String> {
    if recipients::load(context)?.is_empty() {
        return Err("no recipients enrolled (run: dotfile secret enroll <label>)".into());
    }
    sops::require_identity(context, None)?;
    Ok(())
}

pub fn add(context: &Context, args: AddArgs) -> Result<(), String> {
    if args.pkg.is_empty() {
        return Err("--pkg <name> is required".into());
    }
    if !recipients::valid_label(&args.pkg) || args.pkg == "." || args.pkg == ".." {
        return Err("--pkg needs a simple package name".into());
    }
    require_vault(context)?;
    let source = super::expand(context, &args.path);
    let meta = source
        .symlink_metadata()
        .map_err(|e| format!("not a file: {e}"))?;
    if meta.file_type().is_symlink() || !meta.is_file() {
        return Err("refusing to adopt a symlink or non-regular file".into());
    }
    let canonical = fs::canonicalize(&source).map_err(|e| e.to_string())?;
    let home = fs::canonicalize(&context.home).map_err(|e| e.to_string())?;
    if !canonical.starts_with(home) {
        return Err("source must live under $HOME".into());
    }
    let group = if args.macos {
        "macos"
    } else if args.hyprland {
        "linux/hyprland"
    } else if args.kde {
        "linux/kde"
    } else if args.arch {
        "linux/arch"
    } else if args.ubuntu {
        "linux/ubuntu"
    } else if args.linux {
        "linux/common"
    } else {
        "shared"
    };
    let package = context.root.join(group).join(&args.pkg);
    let fresh = !package.exists();
    let name = source
        .file_name()
        .ok_or("source has no filename")?
        .to_string_lossy();
    let destination = package.join(format!("{name}.enc"));
    if destination.symlink_metadata().is_ok() {
        return Err(format!("destination exists: {}", destination.display()));
    }
    let encrypted = sops::encrypt(context, &source, &destination, None)?;
    let mut changes = vec![(destination.clone(), encrypted)];
    let marker = package.join(".secret");
    if !args.no_marker && (args.marker || fresh) && !marker.exists() {
        changes.push((marker, Vec::new()));
    }
    let parent = source.parent().ok_or("source has no parent")?;
    if parent != context.home.join(".config").join(&args.pkg) {
        let target = parent
            .strip_prefix(&context.home)
            .map(|p| format!("~/{}", p.display()))
            .unwrap_or_else(|_| parent.display().to_string());
        let line = format!("{group}/{} = {target}", args.pkg);
        let mut targets = fs::read_to_string(&context.targets_file).map_err(|e| e.to_string())?;
        if !targets.lines().any(|existing| existing.trim() == line) {
            if !targets.is_empty() && !targets.ends_with('\n') {
                targets.push('\n');
            }
            targets.push_str(&line);
            targets.push('\n');
            changes.push((context.targets_file.clone(), targets.into_bytes()));
        }
    }
    let paths = changes.iter().map(|(p, _)| p.clone()).collect::<Vec<_>>();
    recipients::commit(context, changes)?;
    vault::set_mode(&source, 0o600)?;
    recipients::stage(context, &paths)?;
    println!(
        "sealed {} -> {}\nkept {} (0600)",
        source.display(),
        destination.display(),
        source.display()
    );
    Ok(())
}

pub fn status(context: &Context, clean: bool, dry: bool) -> Result<ExitCode, String> {
    let configuration = super::configuration(context)?;
    let entries = vault::plan(&configuration)?;
    let variables = vault::load_variables(context);
    let mut blocked = false;
    if entries.is_empty() {
        println!("no secrets tracked");
    }
    for entry in entries {
        let result = if clean {
            vault::clean(context, &entry, &variables, dry)
        } else {
            vault::inspect(context, &entry, &variables)
        };
        match result {
            Ok(state) => {
                println!(
                    "  {}{state} {}",
                    if dry { "would " } else { "" },
                    entry.destination.display()
                );
                blocked |= matches!(state, "drifted" | "plaintext" | "unresolved" | "failed");
            }
            Err(error) => {
                println!(
                    "  {} {}: {error}",
                    if error.starts_with("unknown:") {
                        "unresolved"
                    } else {
                        "failed"
                    },
                    entry.destination.display()
                );
                blocked = true;
            }
        }
    }
    Ok(if blocked {
        ExitCode::FAILURE
    } else {
        ExitCode::SUCCESS
    })
}

pub fn vars(context: &Context, unused: bool) -> Result<ExitCode, String> {
    let variables = vault::load_variables(context);
    if !variables.ok {
        return Err(variables.note.clone());
    }
    let configuration = super::configuration(context)?;
    let mut entries = vault::plan(&configuration)?;
    for package in &configuration.packages {
        if package.kind == crate::config::PackageKind::System {
            entries.extend(vault::package_entries(&configuration, package, true)?);
        }
    }
    let mut used: BTreeMap<String, BTreeSet<String>> = BTreeMap::new();
    for entry in entries {
        if entry.kind != vault::SecretKind::Template {
            continue;
        }
        for name in
            vault::references(&fs::read_to_string(&entry.source).map_err(|e| e.to_string())?)
        {
            used.entry(name)
                .or_default()
                .insert(entry.destination.display().to_string());
        }
    }
    if variables.values.is_empty() {
        println!("nothing declared in vars.enc.yaml");
    }
    for name in variables.values.keys() {
        if unused && used.contains_key(name) {
            continue;
        }
        println!(
            "  {name}  {}",
            used.get(name)
                .map(|paths| paths.iter().cloned().collect::<Vec<_>>().join(" "))
                .unwrap_or_else(|| "unused".into())
        );
    }
    let missing: Vec<_> = used
        .keys()
        .filter(|name| !variables.values.contains_key(*name))
        .collect();
    for name in &missing {
        println!("  {name} referenced but not defined");
    }
    Ok(if missing.is_empty() {
        ExitCode::SUCCESS
    } else {
        ExitCode::FAILURE
    })
}

pub fn edit(context: &Context, path: &Path) -> Result<ExitCode, String> {
    require_vault(context)?;
    let expanded = super::expand(context, path);
    let declared = context.root.join("vars.enc.yaml");
    let mut entries = if path == Path::new("vars")
        || path == Path::new("vars.enc.yaml")
        || expanded == declared
    {
        vec![vault::SecretEntry {
            source: declared.clone(),
            destination: std::path::PathBuf::new(),
            kind: vault::SecretKind::Encrypted,
        }]
    } else {
        vault::plan(&super::configuration(context)?)?
            .into_iter()
            .filter(|entry| {
                entry.source == expanded
                    || entry.destination == expanded
                    || entry.source.ends_with(path)
                    || entry.destination.ends_with(path)
            })
            .collect()
    };
    if entries.len() != 1 {
        return Err(if entries.is_empty() {
            format!("no tracked secret matches '{}'", path.display())
        } else {
            format!("'{}' matches several secrets", path.display())
        });
    }
    let mut entry = entries.remove(0);
    let fresh = !entry.source.exists();
    let temporary = tempfile::Builder::new()
        .prefix("dotfile-secret-edit-")
        .tempdir()
        .map_err(|e| e.to_string())?;
    let working = temporary
        .path()
        .join(entry.source.file_name().ok_or("secret has no filename")?);
    if fresh {
        if entry.source != declared {
            return Err("nothing to edit".into());
        }
        let seed = temporary.path().join("seed.yaml");
        vault::write_private(&seed, b"{}\n")?;
        let encrypted = sops::encrypt(context, &seed, &entry.source, None)?;
        vault::write_private(&working, &encrypted)?;
    } else {
        vault::write_private(
            &working,
            &vault::read_source(&entry.source, sops::MAX_SECRET_BYTES * 4)?,
        )?;
    }
    let mut command = sops::command(context, None);
    command
        .arg("--config")
        .arg(context.root.join(".sops.yaml"))
        .arg(&working)
        .env("TMPDIR", temporary.path())
        .env("TMP", temporary.path())
        .env("TEMP", temporary.path())
        .stdin(std::process::Stdio::inherit());
    let status = crate::process::status(&mut command)
        .map_err(|e| format!("SOPS editor could not start: {e}"))?;
    if !status.success() && status.code() != Some(200) {
        return Ok(ExitCode::from(status.code().unwrap_or(1) as u8));
    }
    crate::fs::write_generated(
        &entry.source,
        &vault::read_source(&working, sops::MAX_SECRET_BYTES * 4)?,
    )?;
    recipients::stage(context, &[entry.source.clone()])?;
    if status.code() == Some(200) {
        println!("unchanged {}", entry.source.display());
        return Ok(ExitCode::SUCCESS);
    }
    if entry.destination.as_os_str().is_empty() {
        println!("saved {}", entry.source.display());
        return Ok(ExitCode::SUCCESS);
    }
    let variables = vault::load_variables(context);
    let result = vault::materialize(context, &mut entry, &variables, false, true)?;
    println!("{} {}", result.detail, entry.destination.display());
    Ok(if result.blocked {
        ExitCode::FAILURE
    } else {
        ExitCode::SUCCESS
    })
}
